File: script/phys/ammo/Quat.ts
import { impl as ammo } from "./context.js";

export class Quat {
    get x(): number {
        return this.impl.x()
    }

    get y(): number {
        return this.impl.y()
    }

    get z(): number {
        return this.impl.z()
    }

    get w(): number {
        return this.impl.w()
    }

    constructor(public impl = new ammo.btQuaternion(0, 0, 0, 1)) { }

    set(x: number, y: number, z: number, w: number): void {
        this.impl.setValue(x, y, z, w);
    }
}

File: puttyknife/phys/bullet-end/Quat.cpp
#include <phys/Quat.hpp>
#include <LinearMath/btQuaternion.h>

phys::Quat *physQuat_new()
{
    return new btQuaternion();
}

float physQuat_getX(phys::Quat *q)
{
    return q->getX();
}
float physQuat_getY(phys::Quat *q)
{
    return q->getY();
}
float physQuat_getZ(phys::Quat *q)
{
    return q->getZ();
}
float physQuat_getW(phys::Quat *q)
{
    return q->getW();
}

float *physQuat_get(phys::Quat *q)
{
    return *q;
}

void physQuat_set(phys::Quat *q, float x, float y, float z, float w)
{
    q->setValue(x, y, z, w);
}